### ISSUE TITLE
DINOv2 (Vision Transformer) FMHA Fusion Failure in TensorRT 10.8.0.43 with PyTorch 2.7.1 ONNX Export   (with GPU 3080)

### ISSUE THREAD

**Issue by CVKim** [NVIDIA/TensorRT#4537]
I am attempting to optimize a DINOv2 Vision Transformer model (specifically dinov2_vitl14) for deployment on NVIDIA GPUs using TensorRT. My primary goal is to leverage Fused Multi-Head Attention (FMHA) / Flash Attention kernels for significant performance gains, as this is crucial for Transformer optimization.

I have followed the standard procedure of exporting the PyTorch model to ONNX and then converting it to a TensorRT engine using trtexec. However, after building the engine, I consistently observe that the expected _mha or fused_mha_v2 layers, which indicate successful FMHA fusion, are not being generated in the TensorRT build logs (--profilingVerbosity=detailed). This occurs even when explicitly trying to enable FlashAttention during ONNX export using torch.backends.cuda.sdp_kernel(enable_flash=True, enable_math=False, enable_mem_efficient=False).

I have confirmed that for a simpler SelfAttention module without an AttentionMask, FMHA fusion (mha_v2) does occur. My question is specifically about how to achieve FMHA fusion for DINOv2, which involves standard attention mechanisms, and for models that might incorporate attention masks.

Environment
TensorRT Version: 10.8.0.43

NVIDIA GPU: NVIDIA H200 (Hopper architecture - SM90), or relevant GPU for test if different.

NVIDIA Driver Version: (Please fill your exact driver version here, e.g., 550.127.08 for CUDA 12.4 as discussed previously)

CUDA Version: 12.8

cuDNN Version: (Please fill your exact cuDNN version here, e.g., 9.x.x)

Operating System: Linux (via Docker container)

Python Version: 3.11.12

PyTorch Version: 2.7.1

Baremetal or Container: Container (Docker)

Relevant Files
Python ONNX Export Script (export_dinov2.py):

Python

import torch
import argparse
import dinov2 # Assuming dinov2 is installed or accessible

class DinoV2Wrapper(torch.nn.Module):
    def __init__(self, model, use_flash_attention):
        super().__init__()
        self.model = model
        self.use_flash_attention = use_flash_attention

    def forward(self, x):
        # Explicitly force FlashAttention backend if enabled
        with torch.backends.cuda.sdp_kernel(
            enable_flash=self.use_flash_attention,
            enable_math=not self.use_flash_attention,
            enable_mem_efficient=False
        ):
            return self.model(x)

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--disable_flash", action="store_true",
                        help="Disable FlashAttention for ONNX export (use math backend)")
    args = parser.parse_args()

    use_flash = not args.disable_flash

    print("Loading DINOv2 ViT-L/14 model...")
    # Make sure dinov2 is loadable, e.g., via torch.hub.load or direct import path
    model = torch.hub.load("facebookresearch/dinov2", "dinov2_vitl14").eval().cuda().half()
    print("Model loaded.")

    wrapped_model = DinoV2Wrapper(model, use_flash_attention=use_flash)

    dummy_input = torch.randn(1, 3, 224, 224, device="cuda", dtype=torch.float16)
    output_onnx = "dinov2_flash.onnx" if use_flash else "dinov2_baseline.onnx"

    print(f"Exporting model to {output_onnx} with opset 23 and dynamo=True...")

    torch.onnx.export(
        wrapped_model,
        dummy_input,
        output_onnx,
        opset_version=23,
        dynamo=True,  # Enable new ONNX Exporter in Torch 2.7+
        input_names=["images"],
        output_names=["features"],
        dynamic_axes={
            "images": {0: "batch_size"},
            "features": {0: "batch_size"},
        },
        do_constant_folding=True,
        verbose=False
    )

    print(f"✓ Successfully exported to {output_onnx}")
Images/Logs demonstrating the issue:

ONNX Graph (example of non-fused MHA pattern): (Reference: image_d30bfe.png from user's upload). This image visually represents the MatMul -> Add -> Div -> Softmax -> MatMul pattern that is expected to fuse into an MHA layer.

Example Profiling Log (from a different successful test showing mha_v2): (Reference: image_d30bb8.png from user's upload). This log snippet shows that mha_v2 kernels can be generated by TensorRT in certain conditions, but not for my DINOv2 conversion.

(Note: The provided image_d387dc.png seems to be a different part of a graph (likely FFN). You should also clarify if image_d30bfe.png is the specific MHA part of DINOv2's ONNX graph, or just a generic MHA pattern.)

Steps To Reproduce
Save the above Python script as export_dinov2.py.

Run the script to generate ONNX models:

Bash

python export_dinov2.py
# This generates 'dinov2_flash.onnx' (with enable_flash=True)
python export_dinov2.py --disable_flash
# This generates 'dinov2_baseline.onnx' (with enable_flash=False)
Build TensorRT engines using trtexec (e.g., for dinov2_flash.onnx):

Bash

/usr/src/tensorrt/bin/trtexec \
  --onnx=dinov2_flash.onnx \
  --saveEngine=dinov2_flash_fp16.plan \
  --fp16 \
  --builderOptimizationLevel=5 \
  --verbose --profilingVerbosity=detailed \
  --minShapes=images:1x3x224x224 \
  --optShapes=images:1x3x224x224 \
  --maxShapes=images:4x3x336x336 \
  --iterations=300 \
  --avgRuns=100 \
  --dumpProfile \
  --workspace=1000
Review the trtexec output logs (especially with --profilingVerbosity=detailed).

Expected Behavior: A mha or fused_mha_v2 layer should be listed in the TensorRT graph/profile indicating FMHA fusion.

Observed Behavior: No such fused attention kernel is found in the logs for DINOv2 conversion. Instead, the attention operations remain as separate MatMul, Softmax, etc., layers.

Full traceback of errors encountered: N/A (conversion succeeds, but FMHA fusion is not observed, no error traceback).

Have you tried [the latest release]?: Yes, I am using TensorRT 10.8.0.43 (GA release), which should support FMHA.

Can this model run on other frameworks?: Yes, the ONNX models run successfully with ONNX Runtime (polygraphy run dinov2_flash.onnx --onnxrt).

Additional Context / Questions for NVIDIA:

I am seeking guidance on the best practices to ensure FMHA fusion for DINOv2-like Vision Transformer models when converting to TensorRT engines.

Specific requirements/flags: Are there any specific ONNX graph patterns, PyTorch export flags, trtexec flags, or TensorRT Builder configurations for version 10.8.0.43 (or a recommended higher version) that can guarantee FMHA fusion for DINOv2's attention layers?

Troubleshooting for AttentionMasks: My prior testing indicates that attention modules without masks fuse (mha_v2) successfully. How can FMHA be leveraged for self-attention modules with attention masks, which are common in many Transformer architectures?

Performance Comparison Methodology: What is the recommended methodology for conducting a comprehensive performance comparison (Latency, Throughput, Memory Usage) for different precisions (FP32, FP16, BF16, INT8) and dynamic batch/sequence lengths? (e.g., trtexec flags like --avgRuns, --noDataTransfers, nsys profile for specific kernel timings).

Myelin / TensorRT Internal Fusion: Could you clarify how Myelin (or TensorRT's internal graph optimizer) handles attention fusion – is it entirely an Ahead-of-Time process during engine build, or are there runtime aspects for dynamic fusion based on input patterns?

Any insights or recommended strategies would be greatly appreciated.

**If the TensorRT version I plan to use is incompatible, I would appreciate it if you could let me know which version I should test with! My ultimate goal is to perform inference and inspection of DINOv2-family models in C++ using the .trt engine.**

**Comment by cooopermaira** [comment]
Just wondering if this is possibly because your maxShapes is a different input size and your onnx export script doesnt seem to accommodate image size as a dynamic axis.
